easybuild/framework/easyconfig/types.py

```diff
@@ -60,22 +49,47 @@ def is_value_of_type(value, typ_spec):
     @param value: value to check the type of
     @param typ_spec: specific type of dict to check for
     """
-    parent_type = typ_spec[0]
-    extra_reqs = dict(typ_spec[1])
-    # first step: check parent type
-    type_ok = isinstance(value, parent_type)
-    if type_ok:
-        _log.debug("Parent type of value %s matches %s, going in...", value, parent_type)
-        # second step: check additional type requirements
-        if parent_type == dict:
-            extra_req_checkers = {
-                # check whether all keys have allowed types
-                'key_types': lambda val: all([type(el) in extra_reqs['key_types'] for el in val.keys()]),
-                # check whether only allowed keys are used
-                'only_keys': lambda val: set(val.keys()) == set(extra_reqs['only_keys']),
-                # check whether all values have allowed types
-                'value_types': lambda val: all([type(el) in extra_reqs['value_types'] for el in val.values()]),
-            }
+    type_ok = False
+
+    if expected_type in EASY_TYPES:
+        # easy types can be checked using isinstance
+        type_ok = isinstance(val, expected_type)
+        msg = ('FAILED', 'passed')[type_ok]
+        _log.debug("Value type checking of easyconfig parameter '%s' %s: expected '%s', got '%s'",
+                   key, msg, expected_type.__name__, type(val).__name__)
+
+    elif expected_type in CHECKABLE_TYPES:
+        parent_type = expected_type[0]
+        extra_reqs = dict(expected_type[1])
+        # first step: check parent type
+        type_ok = isinstance(value, parent_type)
+        if type_ok:
+            _log.debug("Parent type of value %s matches %s, going in...", value, parent_type)
+            # second step: check additional type requirements
+            if parent_type == dict:
+                extra_req_checkers = {
+                    # check whether all keys have allowed types
+                    'key_types': lambda val:
+                        all([any(is_value_of_type(el, t) for t in extra_reqs['key_types']) for el in val.keys()]),
```

Comment:
maybe we should define small helper functions rather than using lambda's?

it'll certainly help in making this more readable :)